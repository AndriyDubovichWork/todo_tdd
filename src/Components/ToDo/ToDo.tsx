import { Checkbox, FormControlLabel } from '@mui/material';
import React from 'react';

export type TaskType = {
  text: string;
  isCompleted: boolean;
  id: number;
};

type ToDoType = {
  tasks: TaskType[];
  setTasks: React.Dispatch<React.SetStateAction<TaskType[]>>;
};

const ToDo = ({ tasks, setTasks }: ToDoType) => {
  const isCheckedHandler = (id: number, tasks: TaskType[]) => {
    const tasksupdate = tasks.map((task) => {
      if (id === task.id) {
        return {
          id: task.id,
          text: task.text,
          isCompleted: !task.isCompleted,
        };
      }
      return {
        id: task.id,
        text: task.text,
        isCompleted: task.isCompleted,
      };
    });

    setTasks(tasksupdate);
  };
  return (
    <>
      {tasks.map((task: TaskType) => {
        return (
          <FormControlLabel
            key={task.id}
            label={task.text}
            className='column'
            control={
              <Checkbox
                checked={task.isCompleted}
                onChange={() => {
                  isCheckedHandler(task.id, tasks);
                }}
              />
            }
          ></FormControlLabel>
        );
      })}
    </>
  );
};

export default ToDo;
