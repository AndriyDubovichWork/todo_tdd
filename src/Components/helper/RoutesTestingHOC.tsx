import React from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

type RoutesTestingHOCType = {
  children: JSX.Element;
};
const RoutesTestingHOC = ({ children }: RoutesTestingHOCType) => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path='*' element={children} />
      </Routes>
    </BrowserRouter>
  );
};

export default RoutesTestingHOC;
