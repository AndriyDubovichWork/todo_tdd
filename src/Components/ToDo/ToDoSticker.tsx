import React, { Dispatch, SetStateAction, useState } from 'react';
import Draggable from 'react-draggable';
import { Box, Checkbox, TextField } from '@mui/material';
import PushPinIcon from '@mui/icons-material/PushPin';
import CloseIcon from '@mui/icons-material/Close';
import { TaskType } from './ToDo';
import LinesEllipsis from 'react-lines-ellipsis';
type ToDoStickerType = {
  task: TaskType;
  tasks: TaskType[];
  isCheckedHandler: (id: number, tasks: TaskType[]) => void;
  setTasks: any;
  Delete: (
    task: TaskType[],
    setTasks: Dispatch<SetStateAction<TaskType[]>>,
    id: number
  ) => void;
};

const ToDoSticker = ({
  task,
  tasks,
  isCheckedHandler,
  Delete,
  setTasks,
}: ToDoStickerType) => {
  const [isDraggeable, setIsDraggeable] = useState(true);
  const [isTextCheangeable, setIsTextCheangeable] = useState(false);

  const cursor = isDraggeable ? 'grab' : 'auto';
  return (
    <Draggable disabled={!isDraggeable} key={task.id}>
      <Box
        style={{
          padding: '10px',
          backgroundColor: '#cece',
          display: 'inline-block',
          overflowWrap: 'break-word',
          wordWrap: 'break-word',
          hyphens: 'auto',
          cursor: cursor,
          width: 200,
          height: 200,
        }}
      >
        <Checkbox
          data-testid='check-box'
          checked={task.isCompleted}
          onChange={() => {
            isCheckedHandler(task.id, tasks);
          }}
          sx={{ cursor: 'pointer' }}
        />

        {!isDraggeable ? (
          <PushPinIcon
            sx={{ transform: 'rotate(20deg)', cursor: 'pointer' }}
            onClick={() => {
              setIsDraggeable(!isDraggeable);
            }}
          />
        ) : (
          <PushPinIcon
            sx={{ cursor: 'pointer' }}
            onClick={() => {
              setIsDraggeable(!isDraggeable);
            }}
          />
        )}
        <CloseIcon
          sx={{ cursor: 'pointer' }}
          onClick={() => Delete(tasks, setTasks, task.id)}
        />
        {isTextCheangeable ? (
          <TextField label='edit task' multiline rows={5} value={task.text} />
        ) : (
          <LinesEllipsis
            text={task.text}
            onDoubleClick={() => {
              if (isDraggeable) {
                return;
              }
              setIsTextCheangeable(true);
            }}
            maxLine='7'
            ellipsis='...'
            trimRight
            style={{ userSelect: 'none', cursor: cursor }}
            basedOn='letters'
          />
        )}
      </Box>
    </Draggable>
  );
};

export default ToDoSticker;
