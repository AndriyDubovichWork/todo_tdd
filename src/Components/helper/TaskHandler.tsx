import { TaskType } from '../ToDo/ToDo';

export const TaskHandler = (
  tasks: TaskType[],
  setTasks: any,
  task: TaskType,
  newValue: string
) => {
  const newTasks = tasks.map((Innertask: TaskType) => {
    if (Innertask.id === task.id) {
      Innertask.text = newValue;
    }
    return Innertask;
  });
  setTasks(newTasks);
};
