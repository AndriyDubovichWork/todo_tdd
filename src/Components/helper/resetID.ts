import { TaskType } from '../ToDo/ToDo';

export const ResetID = (tasks: TaskType[], setTasks: any) => {
  let copy = tasks.map((task: TaskType, id: number) => {
    return {
      ...task,
      id,
    };
  });
  if (!copy) {
    copy = [];
  }
  setTasks(copy);
};
