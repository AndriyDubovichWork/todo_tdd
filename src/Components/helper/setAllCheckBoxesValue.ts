import { TaskType } from '../ToDo/ToDo';

export const setAllCheckBoxesValue = (
  tasks: TaskType[],
  setTasks: any,
  value: boolean
) => {
  setTasks(
    tasks.map((task: TaskType) => {
      return { ...task, isCompleted: value };
    })
  );
};
