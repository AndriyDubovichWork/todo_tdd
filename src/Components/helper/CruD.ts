import deepcopy from 'deepcopy';
import { TaskType } from '../ToDo/ToDo';
import { ResetID } from './resetID';
import { Dispatch, SetStateAction } from 'react';

export const addToList = (
  tasks: TaskType[],
  task: TaskType,
  setTasks: Dispatch<SetStateAction<TaskType[]>>
) => {
  let copy = deepcopy(tasks);
  task.id = copy.length;
  copy.push(task);
  setTasks(copy);
};

export const DeleteChecked = (tasks: TaskType[], setTasks: any) => {
  let tasksCopy: TaskType[] = [];
  tasks.map((task: TaskType) => {
    if (!task.isCompleted) {
      tasksCopy.push(task);
    }
  });
  ResetID(tasksCopy, setTasks);
};

export const Delete = (tasks: TaskType[], setTasks: Dispatch<SetStateAction<TaskType[]>>, id: number) => {
  let tasksCopy: TaskType[] = [];
  tasks.map((task: TaskType) => {
    if (task.id !== id) {
      tasksCopy.push(task);
    }
  });
  ResetID(tasksCopy, setTasks);
};
