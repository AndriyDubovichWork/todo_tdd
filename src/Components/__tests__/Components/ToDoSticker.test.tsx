import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen } from '@testing-library/react';
import ToDoSticker from './../../ToDo/ToDoSticker';
import { Delete } from './../../helper/CruD';
import { TaskType } from '../../ToDo/ToDo';
afterEach(() => cleanup());
// @ts-nocheck
const task: TaskType = { text: 'hi', id: 0, isCompleted: false };
const taskTwo: TaskType = { text: 'hi', id: 1, isCompleted: true };

let tasks: TaskType[] = [
  { text: 'hi', id: 0, isCompleted: false },
  { text: 'hello', id: 1, isCompleted: true },
];
const setTasks = (newtasks: TaskType[]) => {
  tasks = newtasks;
};
const isCheckedHandler = (id: number, tasks: TaskType[]) => {
  const tasksupdate = tasks.map((task: TaskType) => {
    if (id === task.id) {
      return {
        id: task.id,
        text: task.text,
        isCompleted: !task.isCompleted,
      };
    }
    return {
      id: task.id,
      text: task.text,
      isCompleted: task.isCompleted,
    };
  });

  setTasks(tasksupdate);
};

test('ToDo Sticker have to match snapshot', () => {
  const tree = renderer
    .create(
      <ToDoSticker
        task={task}
        tasks={tasks}
        isCheckedHandler={isCheckedHandler}
        Delete={Delete}
        setTasks={setTasks}
      />,
      {
        createNodeMock: () => document.createElement('textarea'),
      }
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
test('when task completed checkbox is activated', () => {
  render(
    <ToDoSticker
      task={taskTwo}
      tasks={tasks}
      isCheckedHandler={isCheckedHandler}
      Delete={Delete}
      setTasks={setTasks}
    />
  );

  const CheckBox = screen
    .getByTestId('check-box')
    .querySelector('input[type="checkbox"]');

  expect(CheckBox).toHaveProperty('checked', true);
});
