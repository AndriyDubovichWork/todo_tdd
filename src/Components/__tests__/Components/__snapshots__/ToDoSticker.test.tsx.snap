// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo Sticker have to match snapshot 1`] = `
<div
  className="react-draggable MuiBox-root css-0"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  style={
    Object {
      "backgroundColor": "#cece",
      "cursor": "grab",
      "display": "inline-block",
      "height": 200,
      "hyphens": "auto",
      "overflowWrap": "break-word",
      "padding": "10px",
      "transform": "translate(0px,0px)",
      "width": 200,
      "wordWrap": "break-word",
    }
  }
  transform={null}
>
  <span
    className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-bsd8mz-MuiButtonBase-root-MuiCheckbox-root"
    data-testid="check-box"
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={null}
  >
    <input
      checked={false}
      className="PrivateSwitchBase-input css-1m9pwf3"
      data-indeterminate={false}
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="CheckBoxOutlineBlankIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </span>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-lw08pq-MuiSvgIcon-root"
    data-testid="PushPinIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M16 9V4h1c.55 0 1-.45 1-1s-.45-1-1-1H7c-.55 0-1 .45-1 1s.45 1 1 1h1v5c0 1.66-1.34 3-3 3v2h5.97v7l1 1 1-1v-7H19v-2c-1.66 0-3-1.34-3-3z"
      fillRule="evenodd"
    />
  </svg>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-lw08pq-MuiSvgIcon-root"
    data-testid="CloseIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
    />
  </svg>
  <div
    className="LinesEllipsis  "
    onDoubleClick={[Function]}
    style={
      Object {
        "cursor": "grab",
        "userSelect": "none",
      }
    }
  >
    hi
    <wbr />
  </div>
</div>
`;

exports[`when task completed checkbox is activated 1`] = `
<div
  className="react-draggable MuiBox-root css-0"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  style={
    Object {
      "backgroundColor": "#cece",
      "cursor": "grab",
      "display": "inline-block",
      "height": 200,
      "hyphens": "auto",
      "overflowWrap": "break-word",
      "padding": "10px",
      "transform": "translate(0px,0px)",
      "width": 200,
      "wordWrap": "break-word",
    }
  }
  transform={null}
>
  <span
    className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-bsd8mz-MuiButtonBase-root-MuiCheckbox-root"
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={null}
  >
    <input
      checked={false}
      className="PrivateSwitchBase-input css-1m9pwf3"
      data-indeterminate={false}
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="CheckBoxOutlineBlankIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </span>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-lw08pq-MuiSvgIcon-root"
    data-testid="PushPinIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M16 9V4h1c.55 0 1-.45 1-1s-.45-1-1-1H7c-.55 0-1 .45-1 1s.45 1 1 1h1v5c0 1.66-1.34 3-3 3v2h5.97v7l1 1 1-1v-7H19v-2c-1.66 0-3-1.34-3-3z"
      fillRule="evenodd"
    />
  </svg>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-lw08pq-MuiSvgIcon-root"
    data-testid="CloseIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
    />
  </svg>
  <div
    className="LinesEllipsis  "
    onDoubleClick={[Function]}
    style={
      Object {
        "cursor": "grab",
        "userSelect": "none",
      }
    }
  >
    hi
    <wbr />
  </div>
</div>
`;
